### PR TITLE
Refactoring of part of the federated package

### PR DESCRIPTION
**Pull request by lhstrh**
Fixes https://github.com/lf-lang/lingua-franca/issues/1602. Other things done in this PR:
- [x] removal of `LinkedHashMap<String, Object> federationRTIProperties`
- [x] no passing around of `TargetDecl` in cases where only `Target` is needed
- [x] initialization of `TargetConfig` in its constructor
- [x] handling of creating target configurations for federates in a `FedTargetConfig` subclass
- [x] checks for target/OS support not scattered but in a single place
- [x] https://github.com/lf-lang/lingua-franca/issues/1559 (but no answer regarding testing)
- [x] tests

### CODE REVIEW

File: org.lflang/src/org/lflang/TargetConfig.java [hunk 1]
```diff
@@ -52,10 +55,71 @@ public class TargetConfig {
...
     public final Target target;
 
-    public TargetConfig(TargetDecl target) {
+    public TargetConfig(TargetDecl target) { // FIXME: eliminate this constructor if we can
         this.target = Target.fromDecl(target);
     }
 
+    public TargetConfig(Properties args,
+        TargetDecl target,
+        ErrorReporter errorReporter) {
```
Comment (edwardalee) [inline org.lflang/src/org/lflang/TargetConfig.java:64]:
```suggestion
    public TargetConfig(
        Properties args,
        TargetDecl target,
        ErrorReporter errorReporter
    ) {
```

File: org.lflang/src/org/lflang/TargetConfig.java [hunk 2]
```diff
@@ -52,10 +55,71 @@ public class TargetConfig {
 
     public final Target target;
 
-    public TargetConfig(TargetDecl target) {
+    public TargetConfig(TargetDecl target) { // FIXME: eliminate this constructor if we can
         this.target = Target.fromDecl(target);
     }
 
+    public TargetConfig(Properties args,
```
Comment (edwardalee) [inline org.lflang/src/org/lflang/TargetConfig.java:62]: Constructor needs a comment. Are the args command-line args?

File: org.lflang/src/org/lflang/TargetConfig.java [hunk 3]
```diff
@@ -52,10 +55,71 @@ public class TargetConfig {
...
+        if (args.containsKey("runtime-version")) {
+            this.runtimeVersion = args.getProperty("runtime-version");
+        }
+        if (args.containsKey("external-runtime-path")) {
+            this.externalRuntimePath = args.getProperty("external-runtime-path");
+        }
+        if (args.containsKey(TargetProperty.KEEPALIVE.description)) {
+            this.keepalive = Boolean.parseBoolean(
+                args.getProperty(TargetProperty.KEEPALIVE.description));
+        }
```
Comment (edwardalee) [inline org.lflang/src/org/lflang/TargetConfig.java:120]: Could we add a `--tracing` or `-t` command-line option?

File: org.lflang/src/org/lflang/federated/generator/FedFileConfig.java [hunk 4]
```diff
@@ -43,10 +45,11 @@
  */
 public class FedFileConfig extends FileConfig {
 
-    public FedFileConfig(Resource resource, Path srcGenBasePath, boolean useHierarchicalBin) throws IOException {
-        // FIMXE: It is unclear to me that we need this class.
+    public FedFileConfig(
+        Resource resource,
+        Path srcGenBasePath,
+        boolean useHierarchicalBin) throws IOException {
```
Comment (edwardalee) [inline org.lflang/src/org/lflang/federated/generator/FedFileConfig.java:51]:
```suggestion
    public FedFileConfig(
        Resource resource,
        Path srcGenBasePath,
        boolean useHierarchicalBin
    ) throws IOException {
```

File: org.lflang/src/org/lflang/federated/launcher/CBuildConfig.java [hunk 5]
```diff
@@ -84,28 +64,14 @@ String compileCommandForFederate(FederateInstance federate) {
...
-     * directory is the top-level project folder. This is used to create a launcher script
-     * for federates.
-     *
-     * @param federate The federate to execute.
-     */
     @Override
-    protected
-    String executeCommandForRemoteFederate(FederateInstance federate) {
+    public
+    String remoteExecuteCommand() {
```
Comment (edwardalee) [inline org.lflang/src/org/lflang/federated/launcher/CBuildConfig.java:69]:
```suggestion
    public String remoteExecuteCommand() {
```

File: org.lflang/src/org/lflang/federated/launcher/RtiConfig.java [hunk 6]
```diff
@@ -0,0 +1,53 @@
+package org.lflang.federated.launcher;
+
+import java.nio.file.Path;
+
+public class RtiConfig {
```
Comment (edwardalee) [inline org.lflang/src/org/lflang/federated/launcher/RtiConfig.java:5]: Needs a class comment.

**Review by edwardalee — APPROVED** [review]
This looks great!! This is a much needed pass on this code. I've made some really minor suggestions.  This looks to me like it may bring us really close to multilingual federates!  Maybe we just need to add a `mixed` target.

**Comment by lhstrh** [comment]
> This looks to me like it may bring us really close to multilingual federates!

Yes, we're getting closer... Let me jot down a design proposal before going ahead and implementing it. The goal of this PR is just to get to the point of having an understandable handling of target configurations.

**Pull request merged**